### DISCUSSION TITLE
Only use the option of Autofree to manage memory

### DISCUSSION THREAD

**Discussion by Jengro777** [vlang/v#20489] [General]
I read a lot of the author's documents about V, and many of them said that GC brought him surprise, which made me think that V language is still a language that mainly develops GC.

To be honest, I came here for the Autofree function, and I think many people came here for this function.

There are too many languages that use GC. If the V language is dominated by GC, I don't think there will be much development, even if it is a combination of autofree and gc (this combination is uncontrollable).

I think the V language should spend more time on autofree, and provide a complete autofree option when compiling, so that I can know that my program is controllable.

**Comment by medvednikov** [comment]
autofree will be a first class citizen in 0.5.

Like the blog post on autofree explains, and like you mentioned, GC indeed showed a better performance than expected (and better than autofree), but it's clear that autofree is wanted/needed by many.

**Reply by Jengro777** [reply]
Very good, looking forward to autofree becoming a regular official

**Comment by Jengro777** [comment]
I would like to know about when V0.5 will be released?

**Reply by JalonSolov** [reply]
When it is ready.

**Reply by Jengro777** [reply]
When `autofree` is available, is there an option to manage memory using only `autofree`? Some programs cannot tolerate the pauses introduced by garbage collection (GC).


eg:  `v -gc none -prod .`

**Reply by JalonSolov** [reply]
Do you have examples of unacceptable pauses caused by V's GC?

**Reply by Jengro777** [reply]
There are currently no examples, but not using GC is very important for specific scenarios, especially for embedded systems. The V language first needs to support this scenario, and we also don't like to manage memory ourselves; This way, Autofree is the only option

**Reply by Jengro777** [reply]
We cannot tolerate the uncertainty brought by GC, even if the program runs slower, it is acceptable. However, uncertain GC may lead to difficult to troubleshoot memory issues

**Reply by JalonSolov** [reply]
Yes... I'm sure you'll be able to create a trillion objects in an embedded system... They _always_ have terabytes of RAM, right?

If you cut it down to a more believable number... say 1 billion...
```
Program execution time: 26.486s
```
that's what I get on my system.

Now how does that compare to an embedded system, which isn't likely to have 32 GB RAM?  I don't have a clue.  I do know it's still a totally unrealistic "benchmark" for this idea.

If you cut it down to a more realistic 1 million, this is what I get:
```
Program execution time: 27.166ms
```
and again, I have no idea how that compares to an embedded system.

However, that isn't really the question, anyway.  What you need to do is compare gc vs autofree.

So, keeping that same 1 million, and running with `v -gc none -autofree`, I get this:
```
Program execution time: 32.249ms
```
which is actually **_slower_** than the gc version.

So again I ask - do you have an example where V's gc causes a problem?

**Reply by Jengro777** [reply]
There are no cases, but I want to confirm if there will be an option to only use AutoFree? If there is, I will continue to stay and wait. If not, I will choose a more mature go

If I use GC, why don't I use GO instead of V? The biggest reason why I chose to use V is to only use Autofree

**Reply by Jengro777** [reply]
`v -gc none -autofree 
`
Only using AutoFree is essential for ensuring the correct and stable management of memory, rather than the uncertainty brought by GC. Moreover, the Go language performs better in GC.

**Reply by Jengro777** [reply]
`v -gc none -autofree main.v`

```
avey@avey river-x86_64-unknown-linux-gnu % v -gc none -autofree main.v
main.v:13:9: warning: unused variable: `i`
   11 |
   12 |     // Continuously create a large number of temporary objects
   13 |     for i in 0 .. 1_000_000_000_000 {
      |         ^
   14 |         mut obj := MyStruct{data: ['Hello, World!']}
   15 |         // Here we don't actually use obj, it will quickly become garbage
main.v:14:13: warning: unused variable: `obj`
   12 |     // Continuously create a large number of temporary objects
   13 |     for i in 0 .. 1_000_000_000_000 {
   14 |         mut obj := MyStruct{data: ['Hello, World!']}
      |             ~~~
   15 |         // Here we don't actually use obj, it will quickly become garbage
   16 |     }
avey@avey river-x86_64-unknown-linux-gnu %
```


v -gc none -autofree 
In my opinion, it has always been in an unavailable state, and I don't know if there will be separate support for AutoFree in the future. If we don't choose to use the GC language, V is not a better choice

**Reply by Jengro777** [reply]
If Autofree executes slowly, it is because the current compilation and optimizer are not good enough! Continuous optimization can be carried out in the future

**Reply by medvednikov** [reply]
>  I want to confirm if there will be an option to only use AutoFree?

Of course, why wouldn't there be. In fact there already is. `v -autofree`

**Reply by medvednikov** [reply]
Autofree is not finished and not a priority until V 0.6, but it will be finished.

**Comment by Jengro777** [comment]
`v -gc none -autofree run .`
`v -gc none -autofree  .`

`v -gc none -autofree -d veb_livereload watch run .`



```
avey@avey vprod % v -gc none -autofree -d veb_livereload watch run .
2024-10-06 07:49:00.141: "/Applications/v/v" -gc none -autofree -d veb_livereload run . | pid:   11906 | reload cycle:     0
================== C compilation error (from cc): ==============
cc:                       ^
cc: /tmp/v_501/vprod.01J9FHC9KRR2NYPKC2WNJMDFV9.tmp.c:49647:39: error: call to undeclared function 'tp'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
cc: string _arg_expr_split_parse_1_9393 = tp(2, _MOV((StrIntpData[]){{_SLIT0, 0xfe10, {.d_s = line}}, {_SLIT("\n"), 0, { .d_c = 0 }}}));
cc:                                       ^
cc: /tmp/v_501/vprod.01J9FHC9KRR2NYPKC2WNJMDFV9.tmp.c:49647:8: error: initializing 'string' (aka 'struct string') with an expression of incompatible type 'int'
cc: string _arg_expr_split_parse_1_9393 = tp(2, _MOV((StrIntpData[]){{_SLIT0, 0xfe10, {.d_s = line}}, {_SLIT("\n"), 0, { .d_c = 0 }}}));
cc:        ^                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc: /tmp/v_501/vprod.01J9FHC9KRR2NYPKC2WNJMDFV9.tmp.c:49648:5: error: expected expression
cc:                 } else {
cc:                   ^
cc: /tmp/v_501/vprod.01J9FHC9KRR2NYPKC2WNJMDFV9.tmp.c:49653:13: warning: ordered comparison between pointer and integer ('char *(*)(const char *, int)' and 'int')
cc:                 if (index < (int)(lines.len - 1)) {
... (the original output was 755 lines long, and was truncated to 12 lines)
================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error:
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**Reply by JalonSolov** [reply]
The problem is that `-autofree` support isn't complete, yet.  It is still WIP.  It is planned for the near future, but isn't complete yet.

**Comment by Wajinn** [comment]
> "To be honest, I came here for the Autofree function, and I think many people came here for this function."

> "There are too many languages that use GC. If the V language is dominated by GC, I don't think there will be much development, even if it is a combination of autofree and gc (this combination is uncontrollable)."

There is a number of issues, in regards to the statements above: 

1. Many languages using a GC are slower interpreted ones (Python, Lua, Ruby...), where V is a compiled language, thus still fast.
2. Many people have no pressing reason nor a program that must use something other than a GC because of performance. 
3. The majority of users and tasks are going to be just fine with using a GC, and where differences are within seconds or milliseconds.
4. The advantages of using V are beyond just autofree. 
5. Have never seen it indicated that autofree is the ultimate purpose of V.

> "In my opinion, it has always been in an unavailable state, and I don't know if there will be separate support for AutoFree in the future. If we don't choose to use the GC language, V is not a better choice"

I'm not understanding the insistence to make autofree the highest priority, when it's not listed as being that on the website. Don't get me wrong, another and faster form of automatic memory management is a nice feature (add arena allocation in there too), but it's not the only feature. Team V keeps making it clear it's one of many things being or to be worked on. V performs about as well as Go (but you can turn the GC off or choose something else), which are both compiled languages, and where they may have an advantage or disadvantage (to each other) depends on the test used and how the particular code was written.

The V language and other libraries (like VUI or C2V for instance), looks to be of at least equal importance to any particular memory management option. From what I've read around here and the website, looks like various options for manage memory will always be provided. Autofree looks to always be one of various options, versus the only option or the only reason V exists.

**Comment by qpwo** [comment]
I also came for the autofree. I don't mind the developers focusing on GC or taking V somewhere else. It is their project and they set the priorities. But autofree is what got me excited about this.

**Reply by medvednikov** [reply]
There are many people who think like you, so we'll definitely finish autofree and make it on par with GC. Maybe even an optional ownership model as well.

**Reply by medvednikov** [reply]
But yeah like Wajinn eloquently said, GC is fine for 95% of users, and it performs really well with a compiled static and simple language like V. That's why autofree is not a priority atm.

**Reply by Jengro777** [reply]
> There are many people who think like you, so we'll definitely finish autofree and make it on par with GC. Maybe even an optional ownership model as well.

Looking forward to autofree and optional ownership
